File: frontend/src/game/objManager.ts
import { Scene } from "phaser";
import { MapIndex, Position, PressedDir } from "./types";
import { TILE_WIDTH } from "./gameConfig";
import { ManObj } from "./objects/man";
import { BombObj } from "./objects/bomb";
import { FireObj, FireObjConfig } from "./objects/fire";
import { ItemObj } from "./objects/item";
import { PlayerMoveEventPayload, BombExplode, CreateItem, PlayerDie, GenerateBombEvent, RemoveItem, PlayerGetItemEvent, ItemType } from "./event/events";
import { MapManager } from "./mapManager";
import { tranIndexToPosition } from "./utils";
import { ANIMS } from "./sprite_animations/animations";
import { GameMetaData, useGlobalStore } from "../store";
import { BrickObj } from "./objects/brick";
import { SendSelfPlayerMovePayload } from "../websocket/eventMap";



// interface ExplodeData {
//     explodeBombs: BombObj[],
//     ruinBricks: BrickObj[],
//     destroyItems: ItemObj[],
//     showFireConfigs: FireObjConfig[]
// }
export class ObjManager {
    players: ManObj[] = []
    mapManager: MapManager
    bombs: BombObj[] = []
    fires: FireObj[] = []
    ruiningBricks: BrickObj[] = []
    constructor(private scene: Scene, gameMetaData: GameMetaData) {
        this.mapManager = new MapManager(scene, gameMetaData, this.players)
    }
    ///by self
    handleSelfPositionChange(selfManObj: ManObj, pressedDir: PressedDir) {
        const speed = selfManObj.speed
        const prevX = selfManObj.sprite.x
        const prevY = selfManObj.sprite.y
        const currDir = selfManObj.dir
        let targetX = prevX
        let targetY = prevY
        if (pressedDir === "up") targetY -= speed
        if (pressedDir === "down") targetY += speed
        if (pressedDir === "left") targetX -= speed
        if (pressedDir === "right") targetX += speed

        let canMove = this.mapManager.canManMoveByPosition({ posX: targetX, posY: targetY }, selfManObj)
        let finalX = targetX
        let finalY = targetY
        if (!canMove) {
            const threshold = speed
            if (pressedDir === "left" || pressedDir === "right") {
                const offset = prevY % TILE_WIDTH
                if (offset !== 0) {
                    if (offset <= threshold) {
                        finalY = prevY - offset
                        finalX = targetX
                    } else if (TILE_WIDTH - offset <= threshold) {
                        finalY = prevY + (TILE_WIDTH - offset)
                        finalX = targetX
                    }
                    canMove = this.mapManager.canManMoveByPosition({ posX: finalX, posY: finalY }, selfManObj)
                }
            } else if (pressedDir === "up" || pressedDir === "down") {

                const offset = prevX % TILE_WIDTH
                if (offset !== 0) {
                    if (offset <= threshold) {
                        finalX = prevX - offset
                        finalY = targetY
                    } else if (TILE_WIDTH - offset <= threshold) {
                        finalX = prevX + (TILE_WIDTH - offset)
                        finalY = targetY
                    }
                    canMove = this.mapManager.canManMoveByPosition({ posX: finalX, posY: finalY }, selfManObj)
                }
            }

            if (!canMove) {
                if (pressedDir === "up") finalY = Math.ceil(targetY / TILE_WIDTH) * TILE_WIDTH
                else if (pressedDir === "down") finalY = Math.floor(targetY / TILE_WIDTH) * TILE_WIDTH
                else if (pressedDir === "left") finalX = Math.ceil(targetX / TILE_WIDTH) * TILE_WIDTH
                else if (pressedDir === "right") finalX = Math.floor(targetX / TILE_WIDTH) * TILE_WIDTH
                canMove = this.mapManager.canManMoveByPosition({ posX: finalX, posY: finalY }, selfManObj)
            }
        }

        this.handleCanPassBomb(selfManObj, { posY: finalY, posX: finalX })
        const finalDir = pressedDir || currDir
        const finalMoving = !!pressedDir
        // selfManObj.setDir(finalDir)
        // selfManObj.setMoving(finalMoving)
        const eventPayloadToServer = {
            manKey: selfManObj.manSpriteKey,
            dir: finalDir,
            isMoving: finalMoving,
            userId: useGlobalStore.getState().userId,
            newX: canMove ? finalX : prevX,
            newY: canMove ? finalX : prevY,
        } as SendSelfPlayerMovePayload
        // if (canMove) {
        //     eventPayloadToServer.newX = finalX
        //     eventPayloadToServer.newY = finalY
        //     // selfManObj.sprite.setPosition(finalX, finalY)
        // }
        return eventPayloadToServer
    }

    handleCanPassBomb(manObj: ManObj, finalManPos: Position) {
        manObj.canPassBombPosList = manObj.canPassBombPosList.filter(bombP => {
            const dx = Math.abs(bombP.posX - finalManPos.posX)
            const dy = Math.abs(bombP.posY - finalManPos.posY)
            return dx < TILE_WIDTH && dy < TILE_WIDTH
        })
    }

    handleSelfPlaceBomb(playerObj: ManObj) {
        if (playerObj.usedBombNum >= playerObj.bombNum) {
            return null
        }
        const bombIndex = playerObj.getCenterMapIndex()
        const mapTileType = this.mapManager.getMapTileTypeByIndex(bombIndex)
        if (mapTileType === "bomb" || mapTileType === "brick" || mapTileType === "wall") return null
        const newBomb = new BombObj(this.scene, bombIndex, playerObj.bombPower, playerObj.manSpriteKey)
        this.mapManager.setMapTileByIndex(bombIndex, newBomb)
        this.bombs.push(newBomb)
        const pos = tranIndexToPosition(bombIndex)
        playerObj.canPassBombPosList.push(pos)
        playerObj.usedBombNum++
        const payload = {
            manKey: playerObj.manSpriteKey,
            x: bombIndex.indexX,
            y: bombIndex.indexY,
            bombPower: playerObj.bombPower,
            userId: useGlobalStore.getState().userId as number
        }
        return payload
    }

    ///by server
    handlePlayerMoveEventFromServer(playerMoveEventPayload: PlayerMoveEventPayload) {
        // const selfMan = this.players.find(p => p.isSelf)
        // console.log(selfMan, 2222222222)
        // if (selfMan) return
        const { newX, newY, dir, isMoving, manKey } = playerMoveEventPayload
        const targetMan = this.players.find(p => p.manSpriteKey === manKey)
        console.log(playerMoveEventPayload, 2222222222)
        if (!targetMan) {
            return
        }
        targetMan.sprite.setPosition(newX, newY)
        targetMan.setDir(dir)
        targetMan.setMoving(isMoving)
    }
    handleGenerateBombEventFromServer(payload: GenerateBombEvent["payload"]) {
        const index: MapIndex = { indexX: payload.x, indexY: payload.y }
        const targetTileType = this.mapManager.getMapTileTypeByIndex(index)
        if (targetTileType === "empty" || targetTileType === "item") {
            const bomb = new BombObj(this.scene, index, payload.bombPower, payload.manSpriteKey)
            this.mapManager.setMapTileByIndex(index, bomb)
            this.bombs.push(bomb)
        }
    }
    handleCreateItemEventFromServer(payload: CreateItem["payload"]) {
        const index: MapIndex = { indexX: payload.x, indexY: payload.y }
        const item = new ItemObj(this.scene, index, payload.itemType)
        this.mapManager.setMapTileByIndex(index, item)
    }

    handleRemoveItemEventFromServer(payload: RemoveItem["payload"]) {
        const index: MapIndex = { indexX: payload.x, indexY: payload.y }
        const tile = this.mapManager.getMapTileByIndex(index)
        if (tile && tile !== "wall" && tile.type === "item") {
            tile.sprite.destroy()
            this.mapManager.cleanMapTileByIndex(index)
        }
    }

    handlePlayerDieEventFromServer(payload: PlayerDie["payload"][number]) {
        const player = this.players.find(p => p.manSpriteKey === payload.manKey)
        if (!player) return
        player.die()
    }
    handleReomoveFireEventFromServer(payload: { x: number; y: number }) {
        this.fires = this.fires.filter(fire => {
            const cx = fire.centerSprite.x / TILE_WIDTH
            const cy = fire.centerSprite.y / TILE_WIDTH
            if (cx === payload.x && cy === payload.y) {
                fire.destroy()
                return false
            }
            return true
        })
    }
    handleBombExplodeEventFromServer(payload: BombExplode["payload"]) {
        const originIndex: MapIndex = { indexX: payload.x, indexY: payload.y }
        const bombTile = this.mapManager.getMapTileByIndex(originIndex)
        if (bombTile && bombTile !== "wall" && bombTile.type === "bomb") {
            const bombObj = bombTile as BombObj
            const bombOwner = this.players.find(p => p.manSpriteKey === bombObj.manSpriteKey)
            if (bombOwner) bombOwner.usedBombNum = Math.max(0, bombOwner.usedBombNum - 1)
            bombObj.sprite.destroy()
            this.mapManager.cleanMapTileByIndex(originIndex)
            this.bombs = this.bombs.filter(b => {
                const { indexX, indexY } = b.getMapIndex()
                return !(indexX === originIndex.indexX && indexY === originIndex.indexY)
            })
        }

        // Destroy bricks in fire cells (items are handled server-side via createItem/removeItem)
        for (const cell of payload.cells) {
            const cellIndex: MapIndex = { indexX: cell.x, indexY: cell.y }
            const tileType = this.mapManager.getMapTileTypeByIndex(cellIndex)
            const tile = this.mapManager.getMapTileByIndex(cellIndex)
            if (tileType === "brick" && tile && tile !== "wall") {
                const brick = tile as BrickObj
                brick.triggerRuin(() => { })
                this.ruiningBricks.push(brick)
                this.mapManager.setMapTileByIndex(cellIndex, null)
            }
        }

        // Create fire visual by reconstructing spread distances from cells
        const { x, y, cells } = payload
        let up = 0, down = 0, left = 0, right = 0
        for (const cell of cells) {
            if (cell.x === x && cell.y < y) up = Math.max(up, y - cell.y)
            if (cell.x === x && cell.y > y) down = Math.max(down, cell.y - y)
            if (cell.y === y && cell.x < x) left = Math.max(left, x - cell.x)
            if (cell.y === y && cell.x > x) right = Math.max(right, cell.x - x)
        }
        const fire = new FireObj(this.scene, {
            centerX: x, centerY: y,
            verticalStart: up, verticalEnd: down,
            horizontalStart: left, horizontalEnd: right,
        })
        this.fires.push(fire)
    }

    handlePlayerGetItemEventFromServer(payload: PlayerGetItemEvent["payload"]) {
        const player = this.players.find(p => p.manSpriteKey === payload.manKey)
        if (!player) return
        player.eatItem(payload.itemType as ItemType)
    }


}

File: backend2/game/manager/obj_manager.go
package manager

import (
	"math/rand"

	"bomberman/backend2/config"
	"bomberman/backend2/game/objects"
	"bomberman/backend2/types"
)

// ─── Tile types ───────────────────────────────────────────────────────────────

type TileKind int

const (
	TileEmpty TileKind = iota
	TileWall
	TileBomb
	TileBrick
	TileItem
)

type MapCell struct {
	Kind  TileKind
	Bomb  *objects.BombObj
	Brick *objects.BrickObj
	Item  *objects.ItemObj
}

type MapMatrix [][]MapCell

// ─── ExplodeData ──────────────────────────────────────────────────────────────

type ExplodeData struct {
	ExplodeBombPositions  []types.MapIndex
	RuinBrickPositions    []types.MapIndex
	DestroyItemPositions  []types.MapIndex
	ShowFireConfigs       []objects.FireObjConfig
}

// ─── Payloads ─────────────────────────────────────────────────────────────────

type PlaceBombPayload struct {
	ManKey    types.ManSpriteKey
	X         uint32
	Y         uint32
	BombPower uint32
}

type MovePayload struct {
	ManKey   types.ManSpriteKey
	NewX     uint32
	NewY     uint32
	Dir      types.ManDirection
	IsMoving bool
}

// ─── ObjManager ───────────────────────────────────────────────────────────────

type ObjManager struct {
	Players      []*objects.ManObj
	Map          MapMatrix
	Fires        []*objects.FireObj
	RuiningBricks []ruiningBrick
}

type ruiningBrick struct {
	Index       types.MapIndex
	RemainingMs uint32
}

func NewObjManager(mapMatrix MapMatrix, players []*objects.ManObj) *ObjManager {
	return &ObjManager{
		Players: players,
		Map:     mapMatrix,
	}
}

func (o *ObjManager) mapHeight() int { return len(o.Map) }
func (o *ObjManager) mapWidth() int {
	if len(o.Map) == 0 {
		return 0
	}
	return len(o.Map[0])
}

func (o *ObjManager) getTileKind(index types.MapIndex) TileKind {
	if int(index.IndexY) >= o.mapHeight() || int(index.IndexX) >= o.mapWidth() {
		return TileEmpty
	}
	return o.Map[index.IndexY][index.IndexX].Kind
}

func (o *ObjManager) cleanTile(index types.MapIndex) {
	if int(index.IndexY) < o.mapHeight() && int(index.IndexX) < o.mapWidth() {
		o.Map[index.IndexY][index.IndexX] = MapCell{Kind: TileEmpty}
	}
}

func (o *ObjManager) setTile(index types.MapIndex, cell MapCell) {
	if int(index.IndexY) < o.mapHeight() && int(index.IndexX) < o.mapWidth() {
		o.Map[index.IndexY][index.IndexX] = cell
	}
}

// ─── Tick helpers ─────────────────────────────────────────────────────────────

// TickFires decrements fire timers; returns (cx, cy) of expired fires.
func (o *ObjManager) TickFires(passMs uint32) [][2]int {
	var removed [][2]int
	kept := o.Fires[:0]
	for _, f := range o.Fires {
		if f.RemainingMs <= passMs {
			removed = append(removed, [2]int{f.Config.CenterX, f.Config.CenterY})
		} else {
			f.RemainingMs -= passMs
			kept = append(kept, f)
		}
	}
	o.Fires = kept
	return removed
}

func (o *ObjManager) TickRuiningBricks(passMs uint32) {
	kept := o.RuiningBricks[:0]
	for _, rb := range o.RuiningBricks {
		if rb.RemainingMs > passMs {
			rb.RemainingMs -= passMs
			kept = append(kept, rb)
		}
	}
	o.RuiningBricks = kept
}

// GetFireCells returns all tile coords currently on fire.
func (o *ObjManager) GetFireCells() [][2]uint32 {
	var cells [][2]uint32
	for _, f := range o.Fires {
		cx, cy := uint32(f.Config.CenterX), uint32(f.Config.CenterY)
		cells = append(cells, [2]uint32{cx, cy})
		for i := uint32(1); i <= uint32(f.Config.VerticalStart); i++ {
			if cy >= i {
				cells = append(cells, [2]uint32{cx, cy - i})
			}
		}
		for i := uint32(1); i <= uint32(f.Config.VerticalEnd); i++ {
			cells = append(cells, [2]uint32{cx, cy + i})
		}
		for i := uint32(1); i <= uint32(f.Config.HorizontalStart); i++ {
			if cx >= i {
				cells = append(cells, [2]uint32{cx - i, cy})
			}
		}
		for i := uint32(1); i <= uint32(f.Config.HorizontalEnd); i++ {
			cells = append(cells, [2]uint32{cx + i, cy})
		}
	}
	return cells
}

// ─── Explosion ────────────────────────────────────────────────────────────────

func (o *ObjManager) HandleGetExplodeBombData(passMs uint32) ExplodeData {
	result := ExplodeData{}

	// Tick bomb timers, collect expired
	for y, row := range o.Map {
		for x, cell := range row {
			if cell.Kind == TileBomb && cell.Bomb != nil {
				if cell.Bomb.RemainingMs <= passMs {
					cell.Bomb.RemainingMs = 0
					result.ExplodeBombPositions = append(result.ExplodeBombPositions, types.MapIndex{IndexX: uint32(x), IndexY: uint32(y)})
				} else {
					cell.Bomb.RemainingMs -= passMs
					o.Map[y][x].Bomb = cell.Bomb
				}
			}
		}
	}

	// Chain-reaction BFS
	processed := map[[2]uint32]bool{}
	queue := make([]types.MapIndex, len(result.ExplodeBombPositions))
	copy(queue, result.ExplodeBombPositions)

	dirs := [][2]int{{0, -1}, {0, 1}, {-1, 0}, {1, 0}} // up, down, left, right

	for len(queue) > 0 {
		pos := queue[0]
		queue = queue[1:]
		key := [2]uint32{pos.IndexX, pos.IndexY}
		if processed[key] {
			continue
		}
		processed[key] = true

		cell := o.Map[pos.IndexY][pos.IndexX]
		if cell.Kind != TileBomb || cell.Bomb == nil {
			continue
		}
		power := int(cell.Bomb.Power)

		spread := [4]int{}
		for di, d := range dirs {
			for i := 1; i <= power; i++ {
				tx := int(pos.IndexX) + d[0]*i
				ty := int(pos.IndexY) + d[1]*i
				if tx < 0 || ty < 0 || ty >= o.mapHeight() || tx >= o.mapWidth() {
					break
				}
				target := types.MapIndex{IndexX: uint32(tx), IndexY: uint32(ty)}
				kind := o.getTileKind(target)

				if kind == TileWall {
					break
				}
				if i > spread[di] {
					spread[di] = i
				}
				if kind == TileBrick {
					if !indexInList(result.RuinBrickPositions, target) {
						result.RuinBrickPositions = append(result.RuinBrickPositions, target)
					}
					break
				}
				if kind == TileItem {
					if !indexInList(result.DestroyItemPositions, target) {
						result.DestroyItemPositions = append(result.DestroyItemPositions, target)
					}
					break
				}
				if kind == TileBomb {
					tkey := [2]uint32{target.IndexX, target.IndexY}
					if !processed[tkey] && !indexInList(queue, target) {
						result.ExplodeBombPositions = append(result.ExplodeBombPositions, target)
						queue = append(queue, target)
					}
					break
				}
			}
		}

		result.ShowFireConfigs = append(result.ShowFireConfigs, objects.FireObjConfig{
			CenterX:         int(pos.IndexX),
			CenterY:         int(pos.IndexY),
			VerticalStart:   spread[0],
			VerticalEnd:     spread[1],
			HorizontalStart: spread[2],
			HorizontalEnd:   spread[3],
		})
	}

	return result
}

func (o *ObjManager) HandleRenderExplodeByData(data ExplodeData) []SpawnedItem {
	// Spawn fires
	for _, fc := range data.ShowFireConfigs {
		o.Fires = append(o.Fires, objects.NewFireObj(fc))
	}

	// Remove bombs, decrement usedBombNum
	for _, pos := range data.ExplodeBombPositions {
		cell := o.Map[pos.IndexY][pos.IndexX]
		if cell.Kind == TileBomb && cell.Bomb != nil {
			manKey := cell.Bomb.ManKey
			for _, p := range o.Players {
				if p.ManSpriteKey == manKey && p.UsedBombNum > 0 {
					p.UsedBombNum--
				}
			}
		}
		o.cleanTile(pos)
	}

	// Ruin bricks, possibly spawn items
	var spawned []SpawnedItem
	for _, pos := range data.RuinBrickPositions {
		o.cleanTile(pos)
		o.RuiningBricks = append(o.RuiningBricks, ruiningBrick{Index: pos, RemainingMs: config.BrickRuinCountdownMs})

		if rand.Intn(5) <= 2 { // 3/5 chance
			var itemType types.ItemType
			switch rand.Intn(3) {
			case 0:
				itemType = types.ItemSpeed
			case 1:
				itemType = types.ItemMoreBomb
			default:
				itemType = types.ItemFire
			}
			item := objects.NewItemObj(&pos, itemType)
			o.setTile(pos, MapCell{Kind: TileItem, Item: item})
			spawned = append(spawned, SpawnedItem{Index: pos, ItemType: itemType})
		}
	}

	// Destroy items caught in blast
	for _, pos := range data.DestroyItemPositions {
		o.cleanTile(pos)
	}

	return spawned
}

type SpawnedItem struct {
	Index    types.MapIndex
	ItemType types.ItemType
}

// ─── Item pickups ─────────────────────────────────────────────────────────────

type PickupResult struct {
	ManKey   types.ManSpriteKey
	Index    types.MapIndex
	ItemType types.ItemType
}

func (o *ObjManager) HandleItemPickups() []PickupResult {
	var results []PickupResult
	for _, p := range o.Players {
		if !p.IsAlive {
			continue
		}
		idx := p.Base.GetCenterMapIndex()
		cell := o.Map[idx.IndexY][idx.IndexX]
		if cell.Kind == TileItem && cell.Item != nil {
			it := cell.Item.ItemType
			p.EatItem(it)
			o.cleanTile(idx)
			results = append(results, PickupResult{ManKey: p.ManSpriteKey, Index: idx, ItemType: it})
		}
	}
	return results
}

// ─── Bomb placement ───────────────────────────────────────────────────────────

func (o *ObjManager) HandleSelfPlaceBomb(manKey types.ManSpriteKey) *PlaceBombPayload {
	var player *objects.ManObj
	for _, p := range o.Players {
		if p.ManSpriteKey == manKey {
			player = p
			break
		}
	}
	if player == nil || player.UsedBombNum >= player.BombNum {
		return nil
	}
	bombIndex := player.Base.GetCenterMapIndex()
	kind := o.getTileKind(bombIndex)
	if kind == TileBomb || kind == TileBrick || kind == TileWall {
		return nil
	}

	bomb := objects.NewBombObj(&bombIndex, player.BombPower, manKey)
	o.setTile(bombIndex, MapCell{Kind: TileBomb, Bomb: bomb})

	bombPos := types.Position{
		PosX: bombIndex.IndexX * config.TileWidth,
		PosY: bombIndex.IndexY * config.TileWidth,
	}
	player.CanPassBombPosList = append(player.CanPassBombPosList, bombPos)
	player.UsedBombNum++

	return &PlaceBombPayload{
		ManKey:    manKey,
		X:         bombIndex.IndexX,
		Y:         bombIndex.IndexY,
		BombPower: player.BombPower,
	}
}

// ─── Player movement ──────────────────────────────────────────────────────────

func (o *ObjManager) HandlePlayerPositionChange(manKey types.ManSpriteKey, dir types.ManDirection) *MovePayload {
	var player *objects.ManObj
	for _, p := range o.Players {
		if p.ManSpriteKey == manKey {
			player = p
			break
		}
	}
	if player == nil {
		return nil
	}

	prevX, prevY := player.Base.Position.PosX, player.Base.Position.PosY
	speed := player.Speed

	targetX, targetY := prevX, prevY
	switch dir {
	case types.DirUp:
		if targetY >= speed {
			targetY -= speed
		} else {
			targetY = 0
		}
	case types.DirDown:
		targetY += speed
	case types.DirLeft:
		if targetX >= speed {
			targetX -= speed
		} else {
			targetX = 0
		}
	case types.DirRight:
		targetX += speed
	}

	finalX, finalY := targetX, targetY
	canMove := o.canManMoveByPosition(finalX, finalY, player)

	if !canMove {
		tw := config.TileWidth
		switch dir {
		case types.DirLeft, types.DirRight:
			offset := prevY % tw
			if offset != 0 {
				if offset <= speed {
					finalY = prevY - offset
					finalX = targetX
				} else if tw-offset <= speed {
					finalY = prevY + (tw - offset)
					finalX = targetX
				}
				canMove = o.canManMoveByPosition(finalX, finalY, player)
			}
		case types.DirUp, types.DirDown:
			offset := prevX % tw
			if offset != 0 {
				if offset <= speed {
					finalX = prevX - offset
					finalY = targetY
				} else if tw-offset <= speed {
					finalX = prevX + (tw - offset)
					finalY = targetY
				}
				canMove = o.canManMoveByPosition(finalX, finalY, player)
			}
		}

		if !canMove {
			tw := config.TileWidth
			switch dir {
			case types.DirUp:
				finalY = ((targetY + tw - 1) / tw) * tw
			case types.DirDown:
				finalY = (targetY / tw) * tw
			case types.DirLeft:
				finalX = ((targetX + tw - 1) / tw) * tw
			case types.DirRight:
				finalX = (targetX / tw) * tw
			}
			canMove = o.canManMoveByPosition(finalX, finalY, player)
		}
	}

	o.handleCanPassBomb(manKey, finalX, finalY)

	player.Base.Position.PosX = finalX
	player.Base.Position.PosY = finalY
	player.Dir = dir
	player.IsMoving = true

	if canMove {
		return &MovePayload{ManKey: manKey, NewX: finalX, NewY: finalY, Dir: dir, IsMoving: true}
	}
	return nil
}

func (o *ObjManager) handleCanPassBomb(manKey types.ManSpriteKey, finalX, finalY uint32) {
	var player *objects.ManObj
	for _, p := range o.Players {
		if p.ManSpriteKey == manKey {
			player = p
			break
		}
	}
	if player == nil {
		return
	}
	tw := uint64(config.TileWidth)
	kept := player.CanPassBombPosList[:0]
	for _, bp := range player.CanPassBombPosList {
		dx := absDiff(uint64(bp.PosX), uint64(finalX))
		dy := absDiff(uint64(bp.PosY), uint64(finalY))
		if dx < tw && dy < tw {
			kept = append(kept, bp)
		}
	}
	player.CanPassBombPosList = kept
}

func (o *ObjManager) canManMoveByPosition(posX, posY uint32, man *objects.ManObj) bool {
	tw := config.TileWidth
	corners := [][2]uint32{
		{posX, posY},
		{posX + tw - 1, posY},
		{posX, posY + tw - 1},
		{posX + tw - 1, posY + tw - 1},
	}
	for _, c := range corners {
		ix, iy := c[0]/tw, c[1]/tw
		kind := o.getTileKind(types.MapIndex{IndexX: ix, IndexY: iy})
		switch kind {
		case TileWall, TileBrick:
			return false
		case TileBomb:
			// Can pass if player placed it and still overlaps
			canPass := false
			for _, bp := range man.CanPassBombPosList {
				if c[0] >= bp.PosX && c[0] < bp.PosX+tw &&
					c[1] >= bp.PosY && c[1] < bp.PosY+tw {
					canPass = true
					break
				}
			}
			if !canPass {
				return false
			}
		}
	}
	return true
}

func (o *ObjManager) HandlePlayerDie(manKey types.ManSpriteKey) {
	for _, p := range o.Players {
		if p.ManSpriteKey == manKey {
			p.IsAlive = false
			return
		}
	}
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

func indexInList(list interface{}, target types.MapIndex) bool {
	switch l := list.(type) {
	case []types.MapIndex:
		for _, v := range l {
			if v.IndexX == target.IndexX && v.IndexY == target.IndexY {
				return true
			}
		}
	}
	return false
}

func absDiff(a, b uint64) uint64 {
	if a > b {
		return a - b
	}
	return b - a
}
